File: js/stepA_more.js
var types = require('./types');
var reader = require('./reader');
if (typeof module !== 'undefined') {
    var readline = require('./node_readline');
}

// read
function READ(str) {
    return reader.read_str(str);
}

// eval
function is_pair(x) {
    return types.sequential_Q(x) && x.length > 0;
}

function quasiquote(ast) {
    if (!is_pair(ast)) {
        return [types.symbol("quote"), ast];
    } else if (ast[0].value === 'unquote') {
        return ast[1];
    } else if (is_pair(ast[0]) && ast[0][0].value === 'splice-unquote') {
        return [types.symbol("concat"), ast[0][1], quasiquote(ast.slice(1))];
    } else {
        return [types.symbol("cons"), quasiquote(ast[0]), quasiquote(ast.slice(1))];
    }
}

function is_macro_call(ast, env) {
    return types.list_Q(ast) &&
           types.symbol_Q(ast[0]) &&
           env.find(ast[0].value) &&
           env.get(ast[0].value)._ismacro_;
}

function macroexpand(ast, env) {
    while (is_macro_call(ast, env)) {
        var mac = env.get(ast[0]);
        ast = mac.apply(mac, ast.slice(1));
    }
    return ast;
}

function eval_ast(ast, env) {
    if (types.symbol_Q(ast)) {
        return env.get(ast);
    } else if (types.list_Q(ast)) {
        return ast.map(function(a) { return EVAL(a, env); });
    } else if (types.vector_Q(ast)) {
        var v = ast.map(function(a) { return EVAL(a, env); });
        v.__isvector__ = true;
        return v;
    } else if (types.hash_map_Q(ast)) {
        var new_hm = {};
        for (k in ast) {
            new_hm[EVAL(k, env)] = EVAL(ast[k], env);
        }
        return new_hm;
    } else {
        return ast;
    }
}

function _EVAL(ast, env) {
    while (true) {
        //console.log("EVAL:", types._pr_str(ast, true));
        if (!types.list_Q(ast)) {
            return eval_ast(ast, env);
        }
    
        // apply list
        ast = macroexpand(ast, env);
        if (!types.list_Q(ast)) { return ast; }

        var a0 = ast[0], a1 = ast[1], a2 = ast[2], a3 = ast[3];
        switch (a0.value) {
        case "def!":
            var res = EVAL(a2, env);
            return env.set(a1, res);
        case "let*":
            var let_env = new types.Env(env);
            for (var i=0; i < a1.length; i+=2) {
                let_env.set(a1[i].value, EVAL(a1[i+1], let_env));
            }
            return EVAL(a2, let_env);
        case "quote":
            return a1;
        case "quasiquote":
            return EVAL(quasiquote(a1), env);
        case 'defmacro!':
            var func = EVAL(a2, env);
            func._ismacro_ = true;
            return env.set(a1, func);
        case 'macroexpand':
            return macroexpand(a1, env);
        case "js*":
            return eval(a1.toString());
        case ".":
            var el = eval_ast(ast.slice(2), env),
                f = eval(a1.toString());
            return f.apply(f, el);
        case "try*":
            try {
                return EVAL(a1, env);
            } catch (exc) {
                if (a2 && a2[0].value === "catch*") {
                    if (exc instanceof Error) { exc = exc.message; }
                    return EVAL(a2[2], new types.Env(env, [a2[1]], [exc]));
                } else {
                    throw exc;
                }
            }
        case "do":
            eval_ast(ast.slice(1, -1), env);
            ast = ast[ast.length-1];
            break;
        case "if":
            var cond = EVAL(a1, env);
            if (cond === null || cond === false) {
                ast = (typeof a3 !== "undefined") ? a3 : null;
            } else {
                ast = a2;
            }
            break;
        case "fn*":
            return types.new_function(EVAL, a2, env, a1);
        default:
            var el = eval_ast(ast, env), f = el[0], meta = f.__meta__;
            if (meta && meta.exp) {
                ast = meta.exp;
                env = new types.Env(meta.env, meta.params, el.slice(1));
            } else {
                return f.apply(f, el.slice(1));
            }
        }
    }
}

function EVAL(ast, env) {
    var result = _EVAL(ast, env);
    return (typeof result !== "undefined") ? result : null;
}

// print
function PRINT(exp) {
    return types._pr_str(exp, true);
}

// repl
var repl_env = new types.Env();
var rep = function(str) { return PRINT(EVAL(READ(str), repl_env)); };
_ref = function (k,v) { repl_env.set(k, v); }

// Import types functions
for (var n in types.ns) { repl_env.set(n, types.ns[n]); }

_ref('readline', readline.readline)
_ref('read-string', reader.read_str);
_ref('eval', function(ast) { return EVAL(ast, repl_env); });
_ref('slurp', function(f) {
    return require('fs').readFileSync(f, 'utf-8');
});
_ref('slurp-do', function(f) {
    return '(do ' + require('fs').readFileSync(f, 'utf-8') + ')';
});

// Defined using the language itself
rep("(def! not (fn* (a) (if a false true)))");
rep("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");
rep("(defmacro! or (fn* (& xs) (if (empty? xs) nil (if (= 1 (count xs)) (first xs) `(let* (or_FIXME ~(first xs)) (if or_FIXME or_FIXME (or ~@(rest xs))))))))");
rep("(def! load-file (fn* (f) (eval (read-string (slurp-do f)))))");

if (typeof process !== 'undefined' && process.argv.length > 2) {
    for (var i=2; i < process.argv.length; i++) {
        rep('(load-file "' + process.argv[i] + '")');
    }
} else if (typeof require === 'undefined') {
    // Asynchronous browser mode
    readline.rlwrap(function(line) { return rep(line); },
                    function(exc) {
                        if (exc instanceof reader.BlankException) { return; }
                        if (exc.stack) { console.log(exc.stack); } else { console.log(exc); }
                    });
} else if (require.main === module) {
    // Synchronous node.js commandline mode
    while (true) {
        var line = readline.readline("user> ");
        if (line === null) { break; }
        try {
            if (line) { console.log(rep(line)); }
        } catch (exc) {
            if (exc instanceof reader.BlankException) { continue; }
            if (exc.stack) { console.log(exc.stack); } else { console.log(exc); }
        }
    }
} else {
    exports.rep = rep;
}

File: cs/stepA_more.cs
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using Mal;
using MalVal = Mal.types.MalVal;
using MalString = Mal.types.MalString;
using MalSymbol = Mal.types.MalSymbol;
using MalInteger = Mal.types.MalInteger;
using MalList = Mal.types.MalList;
using MalVector = Mal.types.MalVector;
using MalHashMap = Mal.types.MalHashMap;
using MalFunction = Mal.types.MalFunction;
using Env = Mal.env.Env;

namespace Mal {
    class stepA_more {
        // read
        static MalVal READ(string str) {
            return reader.read_str(str);
        }

        // eval
        public static bool is_pair(MalVal x) {
            return x is MalList && ((MalList)x).size() > 0;
        }

        public static MalVal quasiquote(MalVal ast) {
            if (!is_pair(ast)) {
                return new MalList(new MalSymbol("quote"), ast);
            } else {
                MalVal a0 = ((MalList)ast)[0];
                if ((a0 is MalSymbol) &&
                    (((MalSymbol)a0).getName() == "unquote")) {
                    return ((MalList)ast)[1];
                } else if (is_pair(a0)) {
                    MalVal a00 = ((MalList)a0)[0];
                    if ((a00 is MalSymbol) &&
                        (((MalSymbol)a00).getName() == "splice-unquote")) {
                        return new MalList(new MalSymbol("concat"),
                                           ((MalList)a0)[1],
                                           quasiquote(((MalList)ast).rest()));
                    }
                }
                return new MalList(new MalSymbol("cons"),
                                   quasiquote(a0),
                                   quasiquote(((MalList)ast).rest()));
            }
        }

        public static bool is_macro_call(MalVal ast, Env env) {
            if (ast is MalList) {
                MalVal a0 = ((MalList)ast)[0];
                if (a0 is MalSymbol &&
                    env.find(((MalSymbol)a0).getName()) != null) {
                    MalVal mac = env.get(((MalSymbol)a0).getName());
                    if (mac is MalFunction &&
                        ((MalFunction)mac).isMacro()) {
                        return true;
                    }
                }
            }
            return false;
        }

        public static MalVal macroexpand(MalVal ast, Env env) {
            while (is_macro_call(ast, env)) {
                MalSymbol a0 = (MalSymbol)((MalList)ast)[0];
                MalFunction mac = (MalFunction) env.get(a0.getName());
                ast = mac.apply(((MalList)ast).rest());
            }
            return ast;
        }

        static MalVal eval_ast(MalVal ast, Env env) {
            if (ast is MalSymbol) {
                MalSymbol sym = (MalSymbol)ast;
                return env.get(sym.getName());
            } else if (ast is MalList) {
                MalList old_lst = (MalList)ast;
                MalList new_lst = ast.list_Q() ? new MalList()
                                            : (MalList)new MalVector();
                foreach (MalVal mv in old_lst.getValue()) {
                    new_lst.conj_BANG(EVAL(mv, env));
                }
                return new_lst;
            } else if (ast is MalHashMap) {
                var new_dict = new Dictionary<string, MalVal>();
                foreach (var entry in ((MalHashMap)ast).getValue()) {
                    new_dict.Add(entry.Key, EVAL((MalVal)entry.Value, env));
                }
                return new MalHashMap(new_dict);
            } else {
                return ast;
            }
        }


        static MalVal EVAL(MalVal orig_ast, Env env) {
            MalVal a0, a1, a2, res;
            MalList el;

            while (true) {

            //System.out.println("EVAL: " + printer._pr_str(orig_ast, true));
            if (!orig_ast.list_Q()) {
                return eval_ast(orig_ast, env);
            }

            // apply list
            MalVal expanded = macroexpand(orig_ast, env);
            if (!expanded.list_Q()) { return expanded; } 
            MalList ast = (MalList) expanded;

            if (ast.size() == 0) { return ast; }
            a0 = ast[0];

            String a0sym = a0 is MalSymbol ? ((MalSymbol)a0).getName()
                                           : "__<*fn*>__";

            switch (a0sym) {
            case "def!":
                a1 = ast[1];
                a2 = ast[2];
                res = EVAL(a2, env);
                env.set(((MalSymbol)a1).getName(), res);
                return res;
            case "let*":
                a1 = ast[1];
                a2 = ast[2];
                MalSymbol key;
                MalVal val;
                Env let_env = new Env(env);
                for(int i=0; i<((MalList)a1).size(); i+=2) {
                    key = (MalSymbol)((MalList)a1)[i];
                    val = ((MalList)a1)[i+1];
                    let_env.set(key.getName(), EVAL(val, let_env));
                }
                orig_ast = a2;
                env = let_env;
                break;
            case "quote":
                return ast[1];
            case "quasiquote":
                orig_ast = quasiquote(ast[1]);
                break;
            case "defmacro!":
                a1 = ast[1];
                a2 = ast[2];
                res = EVAL(a2, env);
                ((MalFunction)res).setMacro();
                env.set(((MalSymbol)a1).getName(), res);
                return res;
            case "macroexpand":
                a1 = ast[1];
                return macroexpand(a1, env);
            case "try*":
                try {
                    return EVAL(ast[1], env);
                } catch (Exception e) {
                    if (ast.size() > 2) {
                        MalVal exc;
                        a2 = ast[2];
                        MalVal a20 = ((MalList)a2)[0];
                        if (((MalSymbol)a20).getName() == "catch*") {
                            if (e is Mal.types.MalException) {
                                exc = ((Mal.types.MalException)e).getValue();
                            } else {
                                exc = new MalString(e.StackTrace);
                            }
                            return EVAL(((MalList)a2)[2],
                                        new Env(env, ((MalList)a2).slice(1,2),
                                                new MalList(exc)));
                        }
                    }
                    throw e;
                }
            case "do":
                eval_ast(ast.slice(1, ast.size()-1), env);
                orig_ast = ast[ast.size()-1];
                break;
            case "if":
                a1 = ast[1];
                MalVal cond = EVAL(a1, env);
                if (cond == Mal.types.Nil || cond == Mal.types.False) {
                    // eval false slot form
                    if (ast.size() > 3) {
                        orig_ast = ast[3];
                    } else {
                        return Mal.types.Nil;
                    }
                } else {
                    // eval true slot form
                    orig_ast = ast[2];
                }
                break;
            case "fn*":
                MalList a1f = (MalList)ast[1];
                MalVal a2f = ast[2];
                Env cur_env = env;
                return new MalFunction(a2f, env, a1f,
                    args => EVAL(a2f, new Env(cur_env, a1f, args)) );
            default:
                el = (MalList)eval_ast(ast, env);
                var f = (MalFunction)el[0];
                MalVal fnast = f.getAst();
                if (fnast != null) {
                    orig_ast = fnast;
                    env = f.genEnv(el.rest());
                } else {
                    return f.apply(el.rest());
                }
                break;
            }

            }
        }

        // print
        static string PRINT(MalVal exp) {
            return printer._pr_str(exp, true);
        }

        // repl
        static MalVal RE(Env env, string str) {
            return EVAL(READ(str), env);
        }

        static void Main(string[] args) {
            string prompt = "user> ";
            
            // core.cs: defined using C#
            var repl_env = new env.Env(null);
            foreach (var entry in core.ns) {
                repl_env.set(entry.Key, entry.Value);
            }
            repl_env.set("eval", new MalFunction(a => EVAL(a[0], repl_env)));
            MalList _argv = new MalList();
            for (int i=1; i < args.Length; i++) {
                _argv.conj_BANG(new MalString(args[i]));
            }
            repl_env.set("*ARGV*", _argv);

            // core.mal: defined using the language itself
            RE(repl_env, "(def! *host-language* \"c#\")");
            RE(repl_env, "(def! not (fn* (a) (if a false true)))");
            RE(repl_env, "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))");
            RE(repl_env, "(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");
            RE(repl_env, "(defmacro! or (fn* (& xs) (if (empty? xs) nil (if (= 1 (count xs)) (first xs) `(let* (or_FIXME ~(first xs)) (if or_FIXME or_FIXME (or ~@(rest xs))))))))");

            int fileIdx = 0;
            if (args.Length > 0 && args[0] == "--raw") {
                Mal.readline.mode = Mal.readline.Mode.Raw;
                fileIdx = 1;
            }
            if (args.Length > fileIdx) {
                RE(repl_env, "(load-file \"" + args[fileIdx] + "\")");
                return;
            }
            
            // repl loop
            RE(repl_env, "(println (str \"Mal [\" *host-language* \"]\"))");
            while (true) {
                string line;
                try {
                    line = Mal.readline.Readline(prompt);
                    if (line == null) { break; }
                } catch (IOException e) {
                    Console.WriteLine("IOException: " + e.Message);
                    break;
                }
                try {
                    Console.WriteLine(PRINT(RE(repl_env, line)));
                } catch (Mal.types.MalContinue) {
                    continue;
                } catch (Mal.types.MalException e) {
                    Console.WriteLine("Error: " +
                            printer._pr_str(e.getValue(), false));
                    continue;
                } catch (Exception e) {
                    Console.WriteLine("Error: " + e.Message);
                    Console.WriteLine(e.StackTrace);
                    continue;
                }
            }
        }
    }
}
